src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1246,6 +1247,13 @@ internal static LLVMTypeRef GetLLVMTypeForTypeDesc(TypeDesc type)
                                     }
                                     break;
                                 case 4:
+                                    // special case EETypePtrType as its used in __MDArray<System___Canon,2>__Address as a "real" not hidden parameter where it appears as a EETypePtr, but is passed as a generic context from the caller which is i8*
```

Comment:
I think it would be better to fix the type of the hidden argument to be `void*` or `EEType*`. It is what we do everywhere else, e.g. for casting helpers.

Reply:
Unfortunately, LLVM does not allow void *,  https://releases.llvm.org/9.0.0/docs/LangRef.html#pointer-type

> Note that LLVM does not permit pointers to void (void*) nor does it permit pointers to labels (label*). Use i8* instead.

`EEType*` might work but it would complicate the generic lookups, e.g.
```
define i8* @"__GenericLookupFromDict_Internal_CompilerGenerated__Module___InvokeRetOII<System___Canon__System___Canon__System___Canon>_TypeHandle_T2_System___Canon"(i8*, i8*) {
genericHelper:
  %retGep = getelementptr i8, i8* %1, i32 8
  %ptrPtr = bitcast i8* %retGep to i8**
  %paramGep = load i8*, i8** %ptrPtr
  ret i8* %paramGep
}
```
The context is the second parameter, `%1` and the `getelementptr` has been generated with an offset of `8`.  That works as its an `i8*` so the pointer arithmentic is just `%1+8`.  If the second parameter were `EEType *`, then `EEType` is 20 bytes for Wasm so there would need to be an additional cast before the pointer arithmetic, or else it would do `%1 + 8 * 20`

Reply:
We should have a clean consistent contract between the codegen and the runtime helpers. The problem we have is that some runtime helpers take the type arguments as `void*`: https://github.com/dotnet/corert/blob/a8271be78880b2037cc884b86246cc966d9425eb/src/Runtime.Base/src/System/Runtime/TypeCast.cs#L47 . Some take the type arguments as `EETypePtr` (the cases you have hit in this PR). And then there are also a few other outliers like IntPtr.

These differences do not matter for x64 since the calling convention is same for all of them, but they do matter for platforms like WebAssembly.

I think that the right type to standardize on with these helpers is `void*` or `EEType*`. There are pros and cons for each option. My preference would be `EEType*`. Pushing knowledge of `EETypePtr` that some of the helpers take today is unnecessary complication of the contract between codegen and runtime.

The best way to fix this would be to do a pass over all helpers used by codegen (including the Array Address method) and change them to use the same type.

Reply:
We use `IntPtr` as the generic context in the `ValueTypeInstanceMethodWithHiddenParameter` class (that serves a similar purpose).

I wouldn't be opposed to switching to `IntPtr` here too.

In `ValueTypeInstanceMethodWithHiddenParameter` we get to the `IntPtr` in a roundabout way (`Context.GetWellKnownType(WellKnownType.Object).GetKnownField("m_pEEType").FieldType;`. We should just make it `Context.GetWellKnownType(WellKnownType.IntPtr)` so that changing the type of `m_pEEType` doesn't mess up the pointer math in LLVM. There has been a TODO on the m_pEEType field since forever `// TODO: Consider making this EEType* instead of IntPtr.` that we couldn't address in .NET Native because of the architecture of one of the tools used.

Reply:
Heh, so now that GitHub showed me Jan's comment - I'm not opposed to using the types Jan is proposing either. `EETypePtr` is not the right type to use, although it has convent methods on it. The convenient methods can become helpers in the Internal.Runtime.CompilerHelpers namespace if needed (e.g. they `operator==`).

Reply:
Ok, I'll go for `EEType*`.  Shall I do that separately to this PR.  I.e. draft this for now, and create a new PR for the change?  

Do you want me to include the removal of `// TODO: Consider making this EEType* instead of IntPtr` and the change to the signature of `__MDArray<System___Canon,2>__Address` (that's the LLVM mangled name which is all I have to hand but you know what I mean)

Reply:
> Shall I do that separately to this PR

I think that would be easier to review.

> Do you want me to include the removal of // TODO: Consider making this EEType* instead of IntPtr

No, that's for someone who really has nothing better to do :). I don't think it gives us anything besides being "nicer".

Reply:
Thought about the hidden arguments: It may be ok to keep the hidden argument as IntPtr as Michal suggested originally. We do not have the right type for the method-dictionary hidden arguments, and so we will need to use IntPtr for those. If we are using IntPtr for the hidden arguments in this case, why not use IntPtr for the hidden arguments in the other cases too.

> I do that separately to this PR

BTW: If it makes things easier, it would be fine to do just a cleanup of the hidden arguments.

> removal of // TODO: Consider making this EEType* instead of IntPtr

There was not much left to fix this TODO, so I have just done it: #8169

Reply:
Ah thanks, didn't check my email and have done https://github.com/dotnet/corert/pull/8170.

Reply:
It is no longer `EETypePtr` after #8173. 

It would be nice to cleanup all EETypePtr references in this file. I do not really understand why everything still works with them.